sql/src/main/java/io/crate/execution/dml/upsert/TransportShardInsertAction.java

```diff
@@ -0,0 +1,205 @@
+            schemaUpdateClient,
+            tasksService,
+            indicesService,
+            shardStateAction,
+            ShardInsertRequest::new,
+            indexNameExpressionResolver
+        );
+        this.schemas = schemas;
+        this.functions = functions;
+        tasksService.addListener(this);
+    }
+
+    public Context buildContext(ShardInsertRequest request) {
+        String indexName = request.index();
+        DocTableInfo tableInfo = schemas.getTableInfo(RelationName.fromIndexName(indexName), Operation.INSERT);
+        Reference[] insertColumns = request.insertColumns();
+        GeneratedColumns.Validation valueValidation = request.validateConstraints()
+            ? GeneratedColumns.Validation.VALUE_MATCH
+            : GeneratedColumns.Validation.NONE;
+
+        TransactionContext txnCtx = TransactionContext.of(request.sessionSettings());
+        InsertSourceGen insertSourceGen = insertColumns == null
+            ? null
+            : InsertSourceGen.of(txnCtx, functions, tableInfo, indexName, valueValidation, Arrays.asList(insertColumns));
+
+        ReturnValueGen returnValueGen = request.returnValues() == null
+            ? null
+            : new ReturnValueGen(functions, txnCtx, tableInfo, request.returnValues());
+
+        return new Context(insertSourceGen, returnValueGen, request.duplicateKeyAction());
+    }
+
+
+    @Override
+    IndexItemResponse processItem(ShardInsertRequest.Item item, Context context, IndexShard indexShard) throws Exception {
+        VersionConflictEngineException lastException = null;
+        boolean isRetry;
+        for (int retryCount = 0; retryCount < MAX_RETRY_LIMIT; retryCount++) {
+            isRetry = retryCount > 0;
+            BytesReference rawSource;
+            Map<String, Object> source = null;
+            try {
+                // This optimizes for the case where the insert value is already string-based, so we can take directly
+                // the rawSource
+                if (context.insertSourceGen instanceof FromRawInsertSource) {
+                    rawSource = context.insertSourceGen.generateSourceAndCheckConstraintsAsBytesReference(item.insertValues());
+                } else {
+                    source = context.insertSourceGen.generateSourceAndCheckConstraints(item.insertValues());
+                    rawSource = BytesReference.bytes(XContentFactory.jsonBuilder().map(source));
+                }
+            } catch (IOException e) {
+                throw ExceptionsHelper.convertToElastic(e);
+            }
+            item.source(rawSource);
+
+            long version = context.duplicateKeyAction == ShardInsertRequest.DuplicateKeyAction.OVERWRITE ? Versions.MATCH_ANY : Versions.MATCH_DELETED;
+            long seqNo = SequenceNumbers.UNASSIGNED_SEQ_NO;
+            long primaryTerm = SequenceNumbers.UNASSIGNED_PRIMARY_TERM;
+
+            try {
+                Engine.IndexResult indexResult = index(item.id(),
```

Comment:
This index call in the middle of the insert logic makes it tricky to move the around, i could handle it in as parameter or break the whole logic into the subparts which would:

1. Prepare/process source on each item
2. Prepare version/seq/primary for write operation
3. Generate the returnvalues